packages/eslint-plugin/src/rules/no-magic-numbers.ts

```diff
@@ -229,3 +110,111 @@ export default util.createRule<Options, MessageIds>({
     };
   },
 });
+
+/**
+ * Gets the true parent of the literal, handling prefixed numbers (-1 / +1)
+ */
+function getLiteralParent(node: TSESTree.Literal): TSESTree.Node | undefined {
+  if (
+    node.parent?.type === AST_NODE_TYPES.UnaryExpression &&
```

Comment:
hmm, i was thinking what to do with `node.parent?`

parent is always defined in eslint context for nodes (with exception of Program).

Reply:
yeah we should probably fix the types.
I was considering just adding a new type to ts-estree:

```ts
type KnownParent<TNode extends Node, TParent extends Node> = TNode & { parent: TParent };
```

But that type wouldn't doesn't fix cases like this. It would let us express direct parent selectors like `VariableDeclaration > VariableDeclarator`, so it's probably a nice thing to add as a non-breaking feature.

---

I was considering doing this via generics to make it even more explicit. _But this only works if we make it non-optional._
```ts
interface BaseNode<TParent extends Node = Node> {
  // ...
  parent: TParent;
}

interface VariableDeclarator<TParent extends Node = Node> extends BaseNode<TParent> {}

interface Program extends BaseNode {
  parent?: never;
}
```

The problem with making it non-optional is that this complicates the converter code because the converter isn't the one that adds the `parent`, eslint is.

Reply:
q: what do you think mayby about adding d.ts with override for type in eslint-utils or plugin itself?

```
define module ....
```